sockeye/lexical_constraints.py

```diff
@@ -24,6 +24,200 @@
+
+        :param phrase: A list of word IDs to add to this trie node.
+        """
+        return self.final_ids
+
+
+class AvoidState:
+    """
+    Represents the state of a hypothesis in the AvoidTrie.
+    The offset is used to return actual positions in the one-dimensionally-resized array that
+    get set to infinity.
+
+    :param avoid_trie: The trie containing the phrases to avoid.
+    :param state: The current state (defaults to root).
+    """
+    def __init__(self,
+                 avoid_trie: AvoidTrie,
+                 state: AvoidTrie = None) -> None:
+
+        self.root = avoid_trie
+        self.state = state if state else self.root
+
+    def consume(self, word_id: int) -> 'AvoidState':
+        """
+        Consumes a word, and updates the state based on it. Returns new objects on a state change.
+
+        :param word_id: The word that was just generated.
+        """
+        if word_id in self.state.children:
+            return AvoidState(self.root, self.state.step(word_id))
+        elif self.state != self.root:
+            return AvoidState(self.root, self.root)
+        else:
+            return self
+
+    def avoid(self) -> Set[int]:
+        """
+        Returns a set of word IDs that should be avoided. This includes the set of final states from the
+        root node, which are single tokens that must never be generated.
+
+        :return: A set of integers representing words that must not be generated next by this hypothesis.
+        """
+        return self.root.final() | self.state.final()
+
+    def __str__(self) -> str:
+        return str(self.state)
+
+
+class AvoidBatch:
+    """
+    Represents a set of phrasal constraints for all items in the batch.
+    For each hypotheses, there is an AvoidTrie tracking its state.
+
+    :param batch_size: The batch size.
+    :param beam_size: The beam size.
+    :param avoid_list: The list of lists (raw phrasal constraints as IDs, one for each item in the batch).
+    :param global_avoid_trie: A translator-level vocabulary of items to avoid.
+    """
+    def __init__(self,
+                 batch_size: int,
+                 beam_size: int,
+                 avoid_list: Optional[List[RawConstraintList]] = None,
+                 global_avoid_trie: Optional[AvoidTrie] = None) -> None:
+
+        self.global_avoid_states = []  # type: List[AvoidState]
+        self.local_avoid_states = []  # type: List[AvoidState]
+
+        # Store the global trie for each hypothesis
+        if global_avoid_trie is not None:
+            self.global_avoid_states = [AvoidState(global_avoid_trie)] * batch_size * beam_size
+
+        # Store the sentence-level tries for each item in their portions of the beam
+        if avoid_list is not None:
+            for raw_phrases in avoid_list:
+                self.local_avoid_states += [AvoidState(AvoidTrie(raw_phrases))] * beam_size
+
+    def reorder(self, indices: mx.nd.array) -> None:
```

Comment:
to be consistent with the rest of the code these should probably be ```mx.nd.NDArray``` (also below).

Reply:
Ok.